AsyncExpiringLazy/BackgroundMonitor.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Strathweb;
+
+namespace AsyncExpiringLazy
+{
+    class BackgroundMonitor<T> : IDisposable
+    {
+        private readonly Func<Task<ExpirationMetadata<T>>> _factory;
+        private readonly Action<ExpirationMetadata<T>> _onNewItem;
+        private readonly CancellationTokenSource _cts = new CancellationTokenSource();
+        private readonly Task _monitorTask;
+        private int _monitorStarted;
+
+        public BackgroundMonitor(Func<Task<ExpirationMetadata<T>>> factory, Action<ExpirationMetadata<T>> onNewItem)
+        {
+            _factory = factory;
+            _onNewItem = onNewItem;
+            _monitorTask = new Task(async () => await RunMonitor().ConfigureAwait(false));
+        }
+
+        public void StartIfNotStarted()
+        {
+            if (Interlocked.CompareExchange(ref _monitorStarted, 1, 0) == 0)
+            {
+                _monitorTask.Start();
+            }
+        }
+
+        private async Task RunMonitor()
+        {
+            while (!_cts.IsCancellationRequested)
+            {
+                try
+                {
+                    var metadata = await _factory().ConfigureAwait(false);
+                    NotifyAboutNewItem(metadata);
+                    
+                    var checkAgainIn = CalculateNextIteration(metadata);
+                    Debug.WriteLine($"Sleeping the connection monitor for {checkAgainIn.TotalSeconds} seconds");
+                    await Task.Delay(checkAgainIn, _cts.Token).ConfigureAwait(false);
```

Comment:
instead of creating an artificial one tick delay, wouldn't it be better to return `TimeSpan.Zero` and only delay when > 0, otherwise let it go? it adds an extra conditional check but to me it sounds better than a fake minimal delay to avoid it

Reply:
Makes sense, done